web_src/js/index.js

```diff
@@ -2850,7 +2850,12 @@ function initVueComponents() {
           params.set('repo-search-page', `${this.page}`);
         }
 
-        window.history.replaceState({}, '', `?${params.toString()}`);
+        const queryString = params.toString();
+        if (queryString) {
+          window.history.replaceState({}, '', `?${queryString}`);
+        } else {
+          window.history.replaceState({}, '', window.location.pathname);
```

Comment:
Is this branch even necessary?

Reply:
Here's a helper function you can put in `utils.js` that preserves pathname, search and hash:

```js
function setUrl(opts) {
  const url = new URL(window.location.href);
  if ('pathname' in opts) url.pathname = opts.pathname || '/';
  if ('search' in opts) url.search = opts.search || '';
  if ('hash' in opts) url.hash = opts.hash || '';
  const newLocation = String(url).replace(url.origin, '');
  window.history[opts.push ? 'pushState' : 'replaceState'](null, null, newLocation);
}
```

You can then call it like this to set/unset only the `search` part:

```js
setUrl({search: params.toString() || ''});
```

Reply:
The hash is will be out of date and irrelevant if you are pressing these buttons - if anything we should be setting the hash to reference the repo-list table.

Imagine we finally get round to sticking ids on the newsfeed items and you have been sent the newsfeed via an anchor then have scrolled back up to the repo-list - keeping the anchor would mean that on going back you will scroll back to the anchor even though the actual proper context for those query elements is the repo-list.

As the repo-list is the top of the page it's not necessarily needed to set the anchor hash - but - keeping it could be confusing.

Reply:
You can reset the hash too with it. This should be equal to your version:

`````js
setUrl({search: params.toString() || '', hash: ''});
````

Your first case would set a wrong path if gitea is running on a subdirectory as it would set path to parent root directory. I think my function is safer because it only sets the parts that you want to change, not all at once.

Reply:
Actually, it  seems a `replaceState` with `?something` as URL does not kill the path, at least in Firefox. But I would probably not want rely that this behaviour is the same in all browsers.

Reply:
So I think your way is actually fine. I did not relialize browsers were smart enough to do the right thing whit `?search` and `#hash` pseudo urls, but it appears to be working fine because the url standard requires to apply the [base url](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#parse-a-url) so fragments like this will work.

````
> new URL('?search', "http://www.example.com/dir/?search").href
'http://www.example.com/dir/?search'
> new URL((new URL("http://www.example.com/dir/")).pathname, "http://www.example.com/dir/?search").href
'http://www.example.com/dir/'

Reply:
hehe! one of the things that I've recently realised is that it is helpful to run your testing instance through a proxy mounted on a suburl (http://localhost/gitea) as it catches a lot of simple problems.